compiler/rustc_middle/src/ty/diagnostics.rs

```diff
@@ -463,3 +416,78 @@ impl<'v> hir::intravisit::Visitor<'v> for StaticLifetimeVisitor<'v> {
         }
     }
 }
+
+pub struct IsSuggestableVisitor<'tcx, 's> {
+    tcx: TyCtxt<'tcx>,
+    bound_str: Option<&'s str>,
+}
+
+impl<'tcx> TypeVisitor<'tcx> for IsSuggestableVisitor<'tcx, '_> {
+    type BreakTy = ();
+
+    fn visit_ty(&mut self, t: Ty<'tcx>) -> ControlFlow<Self::BreakTy> {
+        match t.kind() {
+            FnDef(..)
+            | Closure(..)
+            | Infer(..)
+            | Generator(..)
+            | GeneratorWitness(..)
+            | Bound(_, _)
+            | Placeholder(_)
+            | Error(_) => {
+                return ControlFlow::Break(());
+            }
+
+            Opaque(did, _) => {
+                let parent = self.tcx.parent(*did);
+                if let hir::def::DefKind::TyAlias | hir::def::DefKind::AssocTy = self.tcx.def_kind(parent)
+                    && let Opaque(parent_did, _) = self.tcx.type_of(parent).kind()
+                    && parent_did == did
+                {
+                    // Okay
+                } else {
+                    return ControlFlow::Break(());
+                }
+            }
+
+            Dynamic(dty, _) => {
+                for pred in *dty {
+                    match pred.skip_binder() {
+                        ExistentialPredicate::Trait(_) | ExistentialPredicate::Projection(_) => {
+                            // Okay
+                        }
+                        _ => return ControlFlow::Break(()),
+                    }
+                }
+            }
+
+            Param(param) => {
+                if let Some(found_bound_str) =
+                    param.name.as_str().strip_prefix("impl ").map(|s| s.trim_start())
```

Comment:
Could we rather work on the `DefId`?
There may be multiple `impl Trait` with the same string name.

Reply:
I would like to do this, but I was having trouble associating a `hir::GenericParam`, `ty::Param`, `DefId`... etc. I don't really know how to convert all of them to each other. Is there anywhere I should look to learn?

Reply:
`def_id == tcx.hir().local_def_id(hir_generic_param.hir_id) == tcx.generics_of(item).type_param(ty_param).def_id`
Does this help?

Reply:
Yup, that probably works. I will try to refactor all of the string stuff to instead use the `DefId` of the generic instead.